C/plugins/south/dht11/plugin.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * FogLAMP south plugin.
+ *
+ * Copyright (c) 2018 OSisoft, LLC
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Amandeep Singh Arora
+ */
+#include <dht11.h>
+#include <plugin_api.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <string>
+#include <logger.h>
+#include <plugin_exception.h>
+#include <config_category.h>
+#include <rapidjson/document.h>
+
+using namespace std;
+
+/**
+ * Default configuration
+ */
+#define CONFIG  "{\"plugin\" : { \"description\" : \"DHT11 plugin\", " \
+                        "\"type\" : \"string\", \"default\" : \"foglamp-dht11\" }, " \
+                "\"asset\" : { \"description\" : \"Asset name\", "\
+                        "\"type\" : \"string\", \"default\" : \"dht11\" }, " \
+                "\"pin\" : { \"description\" : \"RPi pin to which DHT11 is attached\", " \
+                        "\"type\" : \"integer\", \"default\" : \"7\" } "\
+                "}"
+
+
+/**
+ * The DHT11 plugin interface
+ */
+extern "C" {
+
+/**
+ * The plugin information structure
+ */
+static PLUGIN_INFORMATION info = {
+	"DHT11",                  // Name
+	"1.0.0",                  // Version
+	0,    			  // Flags
+	PLUGIN_TYPE_SOUTH,        // Type
+	"1.0.0",                  // Interface version
+	CONFIG                    // Default configuration
+};
+
+/**
+ * Return the information about this plugin
+ */
+PLUGIN_INFORMATION *plugin_info()
+{
+	return &info;
+}
+
+/**
+ * Initialise the plugin, called to get the plugin handle
+ */
+PLUGIN_HANDLE plugin_init(ConfigCategory *config)
+{
+	unsigned int pin;
+
+	if (config->itemExists("pin"))
+        {
+                pin = stoul(config->getValue("pin"), nullptr, 0);
+        }
+
+	DHT11 *dht11= new DHT11(pin);
+
+	if (config->itemExists("asset"))
+                dht11->setAssetName(config->getValue("asset"));
+        else
+                dht11->setAssetName("dht11");
+
+	Logger::getLogger()->info("m_assetName set to %s", dht11->getAssetName());
+
+	return (PLUGIN_HANDLE)dht11;
+}
+
+/**
+ * Start the Async handling for the plugin
+ */
+void plugin_start(PLUGIN_HANDLE *handle)
+{
+}
+
+/**
+ * Poll for a plugin reading
+ */
+Reading plugin_poll(PLUGIN_HANDLE *handle)
+{
+	DHT11 *dht11 = (DHT11*)handle;
```

Comment:
handle is config dict, not sure why its being typecast to DHT11.

Reply:
No, it is not defined to always be this. The Python plugins have tended to do this for connivence, it was never intended that this was the case. The handle is just that, an opaque structure that is used by the plugin for purposes of identification, it has no meaning outside of the plugin.

Reply:
this makes sense to me. It would be great then to align python code the same way as well. 
We may set there to use handle(obj)->setConfig getConfig instead of just making handle the config itself; So that reconfigure callback hook can compare with newConfig.